contrib/gogearbox/gearbox.v1/gearbox.go

```diff
@@ -0,0 +1,129 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+// Package gearbox provides functions to trace the gogearbox/gearbox package (https://github.com/gogearbox/gearbox)
+package gearbox // import "gopkg.in/DataDog/dd-trace-go.v1/contrib/gogearbox/gearbox"
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/telemetry"
+
+	"github.com/gogearbox/gearbox"
+	"github.com/valyala/fasthttp"
+)
+
+const componentName = "gogearbox/gearbox"
+
+func init() {
+	telemetry.LoadIntegration(componentName)
+}
+
+// Middleware returns middleware that will trace incoming requests.
+func Middleware(opts ...Option) func(gctx gearbox.Context) {
+	cfg := newConfig()
+	for _, fn := range opts {
+		fn(cfg)
+	}
+	log.Debug("contrib/gogearbox/gearbox: Configuring Middleware: Service: %#v", cfg)
+	spanOpts := []tracer.StartSpanOption{
+		tracer.ServiceName(cfg.serviceName),
+	}
+	return func(gctx gearbox.Context) {
+		if cfg.ignoreRequest(gctx) {
+			gctx.Next()
+			return
+		}
+		fctx := gctx.Context()
+		spanOpts = defaultSpanTags(spanOpts, fctx)
+		// Create an instance of FasthttpContextCarrier, which embeds *fasthttp.RequestCtx and implements TextMapReader
+		fcc := &FasthttpContextCarrier{
+			reqCtx: fctx,
+		}
+		if sctx, err := tracer.Extract(fcc); err == nil {
+			spanOpts = append(spanOpts, tracer.ChildOf(sctx))
+		}
+		span, _ := tracer.StartSpanFromContext(fctx, "http.request", spanOpts...)
+		defer span.Finish()
+
+		// AFAICT, there is no automatic way to update the fashttp context with the context returned from tracer.StartSpanFromContext
+		// Instead I had to manually add the activeSpanKey onto the fashttp context
+		activeSpanKey := tracer.ContextKey{}
+		fctx.SetUserValue(activeSpanKey, span)
```

Comment:
I read some blog posts that implied it's not a good idea to export custom context key types outside of the package (in this case, the `tracer` package). 
I also found this note in the context pkg: https://pkg.go.dev/context#WithValue
> exported context key variables' static type should be a pointer or interface.
So maybe exporting ContextKey is fine, but the type should change?

The tracer package has a function `ContextWithSpan` that essentially does what I do here, but it returns a context.Context with underlying type `*context.valueCtx`. fasthttp's [RequestCtx](https://pkg.go.dev/github.com/valyala/fasthttp#RequestCtx) implements the context.Context interface, but I couldn't cast the return value from `ContextWithSpan` back to *fasthttp.RequestCtx.